File: dcp/173/173.py
#!/usr/bin/env python3

def flatten(m, acc=None, prefix=""):
    acc = acc or {}
    for k, v in m.items():
        if isinstance(v, dict):
            flatten(v, acc, prefix+k+".")
        else:
            acc[prefix+k]=v
    return acc

def main():
    m = {"key": 3, "foo": {"a": 5, "bar": {"baz": 8}}}
    print(repr(flatten(m)))

if __name__ == '__main__':
    main()

File: dcp/173/173.go
/*

Good morning! Here's your coding interview problem for today.

This problem was asked by Stripe.

Write a function to flatten a nested dictionary. Namespace the keys with a period.

For example, given the following dictionary:

{
    "key": 3,
    "foo": {
        "a": 5,
        "bar": {
            "baz": 8
        }
    }
}
it should become:

{
    "key": 3,
    "foo.a": 5,
    "foo.bar.baz": 8
}
You can assume keys do not contain dots in them, i.e. no clobbering will occur.

*/
package dcp173

import "fmt"

type M map[string]interface{}

func flatten(m M) map[string]int {
	acc := make(map[string]int)
	var helper func(prefix string, m M)
	helper = func(prefix string, m M) {
		for key, val := range m {
			switch val := val.(type) {
			case M:
				helper(prefix+key+".", val)
			case int:
				acc[prefix+key] = val
			default:
				panic(fmt.Sprintf("flatten: unexpected value type: %T", val))
			}
		}
	}
	helper("", m)
	return acc
}
